File: src/main/focusGuard.ts
import { exec } from 'child_process'
import { TimerEngine } from './timer'
import { closeBlockerWindow, showBlockerWindow } from './windows'

// Frontmost app name + its front-window title.
const FRONT_SCRIPT = `tell application "System Events"
  set p to first process whose frontmost is true
  set appName to name of p
  set winTitle to ""
  try
    set winTitle to name of front window of p
  end try
  return appName & "||" & winTitle
end tell`

// Chromium-family browsers expose "active tab"; Safari uses "current tab".
const CHROMIUM = [
  'Arc', 'Dia', 'Google Chrome', 'Google Chrome Canary', 'Brave Browser',
  'Microsoft Edge', 'Vivaldi', 'Chromium', 'Opera'
]
function isBrowser(app: string): boolean {
  return CHROMIUM.includes(app) || app === 'Safari' || app === 'Safari Technology Preview'
}
function tabAccessor(app: string): string {
  return app === 'Safari' || app === 'Safari Technology Preview' ? 'current tab' : 'active tab'
}
function urlScript(app: string): string | null {
  if (!isBrowser(app)) return null
  return `tell application "${app}" to get URL of ${tabAccessor(app)} of front window`
}

function run(script: string): Promise<{ out: string; err: string }> {
  return new Promise((res) => {
    exec(`osascript -e '${script.replace(/'/g, "'\\''")}'`, { timeout: 1500 }, (e, out, se) =>
      res({ out: (out || '').trim(), err: e ? se || e.message : '' })
    )
  })
}

export interface Detection {
  app: string
  title: string
  url: string
  error: string
}

/** What is the user looking at right now? Reads the active app + browser URL. */
export async function detectFront(): Promise<Detection> {
  const front = await run(FRONT_SCRIPT)
  if (front.err) return { app: '', title: '', url: '', error: front.err }
  const [app = '', title = ''] = front.out.split('||')
  let url = ''
  const us = urlScript(app)
  if (us) {
    const r = await run(us)
    if (!r.err) url = r.out
  }
  return { app, title, url, error: '' }
}

/** Tokens to match for a blocklist entry: "https://youtube.com/x" -> ["youtube.com","youtube"]. */
function tokens(entry: string): string[] {
  const e = entry.toLowerCase().trim().replace(/^https?:\/\//, '').replace(/^www\./, '')
  const host = e.split('/')[0]
  const main = host.split('.')[0]
  return [...new Set([host, main, e])].filter((t) => t.length >= 3)
}

/** Which blocklist entry (if any) matches what's on screen. */
export function matchBlock(d: Detection, blockList: string[]): string | null {
  const a = d.app.toLowerCase()
  if (a.includes('pomodoro') || a.includes('electron')) return null
  const hay = `${d.app} ${d.title} ${d.url}`.toLowerCase()
  return blockList.find((e) => tokens(e).some((t) => hay.includes(t))) ?? null
}

function matchesUrl(url: string, blockList: string[]): string | null {
  const u = url.toLowerCase()
  return blockList.find((e) => tokens(e).some((t) => u.includes(t))) ?? null
}

/**
 * The real enforcement: for a browser, find every window whose active tab is a
 * blocked site and navigate it to about:blank. Returns the matched entry (if
 * any) so the caller can also show the overlay. Browser-agnostic, no root.
 */
async function enforceBrowser(app: string, blockList: string[]): Promise<string | null> {
  const acc = tabAccessor(app)
  const enumScript = `tell application "${app}"
set out to ""
set i to 0
repeat with w in windows
set i to i + 1
try
set out to out & i & "||" & (URL of ${acc} of w) & "\n"
end try
end repeat
return out
end tell`
  const r = await run(enumScript)
  if (r.err) return null

  let matched: string | null = null
  const hitIdx: number[] = []
  for (const ln of r.out.split('\n').map((l) => l.trim()).filter(Boolean)) {
    const [idx, url = ''] = ln.split('||')
    const hit = matchesUrl(url, blockList)
    if (hit) {
      hitIdx.push(parseInt(idx, 10))
      matched = hit
    }
  }
  if (hitIdx.length) {
    const redir =
      `tell application "${app}"\n` +
      hitIdx
        .map((i) => `try\nset URL of ${acc} of window ${i} to "about:blank"\nend try`)
        .join('\n') +
      `\nend tell`
    await run(redir)
  }
  return matched
}

let engineRef: TimerEngine | null = null
let timer: ReturnType<typeof setInterval> | null = null
let blockedNow: string | null = null
let snoozeUntil = 0

function guardActive(): boolean {
  if (!engineRef) return false
  const s = engineRef.getState()
  return (
    s.settings.strictMode &&
    s.phase === 'focus' &&
    s.status === 'running' &&
    s.settings.blockList.length > 0
  )
}

async function poll(): Promise<void> {
  if (!engineRef) return
  if (!guardActive()) {
    stopGuard()
    return
  }
  if (Date.now() < snoozeUntil) return

  const d = await detectFront()
  if (process.env['POMO_GUARD_DEBUG'])
    console.log('[guard] app=%j title=%j url=%j err=%j', d.app, d.title, d.url, d.error)
  if (d.error || !d.app) return // can't see (no permission) — fail open
  // Ignore our own windows (e.g. the overlay itself is frontmost) — never use
  // that as a reason to close the overlay, or it flickers and lets the page show.
  const a = d.app.toLowerCase()
  if (a.includes('pomodoro') || a.includes('electron')) return

  const blockList = engineRef.getState().settings.blockList
  // For browsers, actively navigate any blocked tab away (the real block).
  let hit: string | null = null
  if (isBrowser(d.app)) hit = await enforceBrowser(d.app, blockList)
  // Fallback for native apps (e.g. "Slack") or when no tab matched.
  if (!hit) hit = matchBlock(d, blockList)

  if (process.env['POMO_GUARD_DEBUG']) console.log('[guard] hit=%j', hit)

  if (hit) {
    blockedNow = hit
    showBlockerWindow(hit)
  } else if (blockedNow) {
    blockedNow = null
    closeBlockerWindow()
  }
}

export function setupFocusGuard(engine: TimerEngine): void {
  // The active-window detection + tab redirect is macOS-only (AppleScript).
  if (process.platform !== 'darwin') return
  engineRef = engine
  engine.onChange(() => {
    if (guardActive() && !timer) {
      timer = setInterval(poll, 1200)
      void poll()
    } else if (!guardActive() && timer) {
      stopGuard()
    }
  })
}

/** Run one detection immediately (used by the Settings "Test" button). */
export async function testGuard(): Promise<Detection & { blocked: string | null }> {
  if (process.platform !== 'darwin') {
    return { app: '', title: '', url: '', error: 'unsupported-os', blocked: null }
  }
  const d = await detectFront()
  const blocked = engineRef ? matchBlock(d, engineRef.getState().settings.blockList) : null
  return { ...d, blocked }
}

export function snoozeGuard(seconds = 6): void {
  snoozeUntil = Date.now() + seconds * 1000
  blockedNow = null
  closeBlockerWindow()
}

export function stopGuard(): void {
  if (timer) {
    clearInterval(timer)
    timer = null
  }
  blockedNow = null
  closeBlockerWindow()
}

File: src-tauri/src/focus_guard.rs
//! macOS strict-mode site blocker — a direct port of `focusGuard.ts`. Pure AppleScript
//! via `osascript`, so it needs **no Node and no Tauri sidecar**. Gated to macOS; on
//! other platforms every fn is a no-op.
//!
//! Behaviour (incl. the v1.0.6 fix): the full-screen overlay is shown ONLY when the
//! *foreground* browser tab is a blocked site, while blocked tabs in background windows
//! are silently redirected to about:blank.
use std::sync::atomic::{AtomicU64, Ordering};
use tauri::{AppHandle, Manager};

static SNOOZE_UNTIL: AtomicU64 = AtomicU64::new(0);

fn now_ms() -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    // Never unwrap: this is on the per-tick path and the release profile aborts on
    // panic, which would kill the timer session. A clock before the epoch just maps
    // to 0 (snooze still works correctly relative to a sane clock).
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_millis() as u64).unwrap_or(0)
}

pub fn snooze(seconds: u64) {
    SNOOZE_UNTIL.store(now_ms() + seconds * 1000, Ordering::Relaxed);
}

/// Called every tick from the timer loop. Runs detection only while strict-focus is on.
pub fn maybe_sync(app: &AppHandle) {
    #[cfg(target_os = "macos")]
    {
        use crate::timer::AppState;
        let (active, block_list) = {
            let s = app.state::<AppState>();
            let e = s.engine.lock().unwrap();
            let on = e.settings.strict_mode
                && matches!(e.phase, crate::model::Phase::Focus)
                && matches!(e.status, crate::model::Status::Running)
                && !e.settings.block_list.is_empty();
            (on, e.settings.block_list.clone())
        };
        if !active {
            crate::windows::hide_blocker(app);
            return;
        }
        if now_ms() < SNOOZE_UNTIL.load(Ordering::Relaxed) {
            crate::windows::hide_blocker(app);
            return;
        }
        let det = detect_front();
        // When our own overlay grabbed focus, Petomato is the frontmost app. Don't
        // treat that as "user navigated away" and hide the overlay — that would start
        // a show/hide/focus-steal flicker loop against the still-blocked browser tab.
        // Hold the overlay until a real, non-blocked foreground app appears.
        if det.app.to_lowercase().contains("petomato") && crate::windows::blocker_visible(app) {
            return;
        }
        // foreground app/url; redirect background blocked tabs silently, overlay only on
        // the foreground tab (see focusGuard.ts poll()).
        if let Some(hit) = enforce_and_match(&det, &block_list) {
            crate::windows::show_blocker(app, &hit);
        } else {
            crate::windows::hide_blocker(app);
        }
        let _ = app; // used on macOS only
    }
    #[cfg(not(target_os = "macos"))]
    {
        let _ = app; // site blocking is macOS-only (AppleScript)
    }
}

/// One-shot detection for the Settings "Test focus shield" button.
pub fn test_once() -> serde_json::Value {
    #[cfg(target_os = "macos")]
    {
        let d = detect_front();
        serde_json::json!({ "app": d.app, "title": d.title, "url": d.url, "error": d.error, "blocked": serde_json::Value::Null })
    }
    #[cfg(not(target_os = "macos"))]
    {
        serde_json::json!({ "app": "", "title": "", "url": "", "error": "unsupported-os", "blocked": null })
    }
}

#[cfg(target_os = "macos")]
struct Detection { app: String, title: String, url: String, error: String }

#[cfg(target_os = "macos")]
fn osascript(script: &str) -> Result<String, String> {
    use std::process::Command;
    let out = Command::new("osascript").arg("-e").arg(script).output().map_err(|e| e.to_string())?;
    if out.status.success() {
        Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
    } else {
        Err(String::from_utf8_lossy(&out.stderr).trim().to_string())
    }
}

#[cfg(target_os = "macos")]
fn detect_front() -> Detection {
    const FRONT: &str = r#"tell application "System Events"
  set p to first process whose frontmost is true
  set appName to name of p
  set winTitle to ""
  try
    set winTitle to name of front window of p
  end try
  return appName & "||" & winTitle
end tell"#;
    match osascript(FRONT) {
        Err(e) => Detection { app: String::new(), title: String::new(), url: String::new(), error: e },
        Ok(s) => {
            let mut it = s.splitn(2, "||");
            let app = it.next().unwrap_or("").to_string();
            let title = it.next().unwrap_or("").to_string();
            let url = if is_browser(&app) {
                let acc = tab_accessor(&app);
                osascript(&format!("tell application \"{app}\" to get URL of {acc} of front window")).unwrap_or_default()
            } else { String::new() };
            Detection { app, title, url, error: String::new() }
        }
    }
}

#[cfg(target_os = "macos")]
fn is_browser(app: &str) -> bool {
    const C: [&str; 9] = ["Arc","Dia","Google Chrome","Google Chrome Canary","Brave Browser","Microsoft Edge","Vivaldi","Chromium","Opera"];
    C.contains(&app) || app == "Safari" || app == "Safari Technology Preview"
}
#[cfg(target_os = "macos")]
fn tab_accessor(app: &str) -> &'static str {
    if app.starts_with("Safari") { "current tab" } else { "active tab" }
}

/// Tokens for a blocklist entry, e.g. "https://youtube.com/x" → ["youtube.com","youtube"].
#[cfg(target_os = "macos")]
fn matches(hay: &str, block_list: &[String]) -> Option<String> {
    let h = hay.to_lowercase();
    block_list.iter().find(|e| {
        let e = e.to_lowercase();
        let e = e.trim_start_matches("https://").trim_start_matches("http://").trim_start_matches("www.");
        let host = e.split('/').next().unwrap_or(e);
        let main = host.split('.').next().unwrap_or(host);
        [host, main, e].iter().any(|t| t.len() >= 3 && h.contains(t))
    }).cloned()
}

#[cfg(target_os = "macos")]
fn enforce_and_match(d: &Detection, block_list: &[String]) -> Option<String> {
    if d.error.is_empty() == false || d.app.is_empty() { return None; }
    let al = d.app.to_lowercase();
    if al.contains("petomato") { return None; }
    if is_browser(&d.app) {
        // silently redirect every blocked tab across windows (real enforcement)
        let acc = tab_accessor(&d.app);
        let enum_script = format!(
            "tell application \"{app}\"\nset out to \"\"\nset i to 0\nrepeat with w in windows\nset i to i + 1\ntry\nset out to out & i & \"||\" & (URL of {acc} of w) & linefeed\nend try\nend repeat\nreturn out\nend tell",
            app = d.app, acc = acc);
        if let Ok(list) = osascript(&enum_script) {
            let hits: Vec<&str> = list.lines().filter_map(|ln| {
                let mut it = ln.splitn(2, "||");
                let idx = it.next()?; let url = it.next().unwrap_or("");
                if matches(url, block_list).is_some() { Some(idx) } else { None }
            }).collect();
            for idx in hits {
                let _ = osascript(&format!("tell application \"{}\" to set URL of {} of window {} to \"about:blank\"", d.app, acc, idx));
            }
        }
        // overlay only when the FOREGROUND tab is blocked
        matches(&d.url, block_list)
    } else {
        // native app in front (e.g. Slack desktop)
        matches(&format!("{} {}", d.app, d.title), block_list)
    }
}
